### PR TITLE
feat: Expose repository DB credentials as a Git credential helper, for #6591

### LINKED ISSUE #6591: Support use of Git repository credentials during Helm dependency update
# Summary

Support should be added for using private Git repositories as the repositories for Helm sub-chart dependencies.  This is currently not available.

# Motivation

Say you have an umbrella chart containing many inter-connected sub-applications, each managed in their own private Git repository, with their Helm chart source alongside them.  You want to use Git for versioning your Charts, rather than a chart museum, so you specify the sub-chart repositories using the [`helm-git` plugin](https://github.com/aslafy-z/helm-git) and the `git+https://` downloader plugin it provides.  Developers who wish to deploy the umbrella chart locally specify their own credentials for the private sub-application repositories via the `credentials.helper store` git config and their `~/.git-credentials` file.

You now want to deploy the Umbrella chart on production with ArgoCD.  You do the equivalent of setting up the `git config credentials.helper store`, i.e. setting up each sub-application Git repository with appropriate credentials as a separate HTTPS Git repository in Settings > Repositories.

You try to create an Application for the Umbrella chart, but you get the following error:
```
`helm dependency build` failed exit status 1: fatal: could not read Username for 'https://github.com': No such device or address
fatal: could not read Username for 'https://github.com': No such device or address
Error in plugin 'helm-git': Unable to fetch remote. Check your Git url.
```

# Proposal

Credentials for all configured Git repositories should be available at templating time, presumably via a `git config credentials.helper store` mechanism.

# Alternative for my usecase

If ArgoCD Helm repository specifications were to support Helm downloader plugin protocols, then these sub-application Git repositories could be specified as Helm repositories directly using the `git+https://` protocol, and the Umbrella chart could then refer to them by reference (e.g. "@sub-application").  However, Helm downloader plugins are not currently utilised when trying to create connections to Helm repositories via the Argo CD UI.

This could be implemented by using the `helm repo add` command (or equivalent Go code) under the hood to verify the settings.

# Caveat

This is all predicated on having the appropriate Helm plugins installed into the image used for the Argo CD server (etc.).

### PR DESCRIPTION
**Pull request by scalen**
Prior to this change, Argo CD repository credentials were available to
Git activity internalised by the Argo CD binary.  However, any
extensions that escaped the scope of the binary (e.g. config management
tools, Helm plugins, etc) and utilised Git were left unable to access
private repositories, even those whose credentials were available to
Argo CD.

This change implements the proposal in and closes #6591. It introduces
an entrypoint to the ArgoCD binary that accepts a repository query as
per the Git Credential Helper specification and tries to match it against
a repository in the Argo CD DB when run; If a match is found, the
credentials are printed to the stdout, as expected of a Git credential
helper.  Furthermore, this new entrypoint is globally configured as Git's
primary credential helper, and Git is configured to useHttpPaths when
querying for credentials.

No documentation has been or needs to be updated, as this is very
much an under-the-hood feature, supporting edge-cases that could
sensibly have been expected to Just Work™️, but didn't.

Checklist:

* [ ] Either I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Review by alexmt — CHANGES_REQUESTED** [review]
Thank you for PR @scalen. 

I agree this is useful functionality. We have a similar command `argocd-util cluster kubeconfig` to extract k8s creds. Instead of adding new binary I suggest to follow the same approach and add new command to `argocd-util`. E.g. `argocd-util repo creds`

File: gitcredentialhelper/gitcredentialhelper.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
+package gitcredentialhelper
```
Comment (alexmt) [inline gitcredentialhelper/gitcredentialhelper.go:1]: Can we please avoid adding one more top level package. Instead I would move this package to `cmd/argocd-util/commands/repocreds`

**Review by alexmt — CHANGES_REQUESTED** [review]

**Comment by jannfis** [comment]
After reading the corresponding issue, I'm not sure whether this will work out as expected. 

Is the goal to have a config management plugin run this credential helper to get the Git credentials in order to access a given repository?

**Comment by scalen** [comment]
> Is the goal to have a config management plugin run this credential helper to get the Git credentials in order to access a given repository?

The goal is to be able to specify repository credentials as normal in ArgoCD, then have a config management plugin be able to pull in secondary repos with just `git clone <repo URL>`, as opposed to having to include credentials in plain text.

This command would not be expected to be called directly, and should probably not be documented. Rather, Git is configured as part of this PR to access this endpoint whenever a repository it tries to access requests credentials.

File: gitcredentialhelper/gitcredentialhelper.go (hunk 1 above)
Comment (scalen) [reply gitcredentialhelper/gitcredentialhelper.go]: Thanks for this suggestion.  I agree: if this isn't to be a top-level command then it doesn't warrant a top-level package.

However, there are no other code directories in `cmd/argocd-util/commands`, so adding one for this feels weird...

I have instead moved it into the `util/git` package.  I'm not 100% that it belongs here, though, or whether it should be in its own `util/` package.  @alexmt, what do you think?

**Comment by jannfis** [comment]
> The goal is to be able to specify repository credentials as normal in ArgoCD, then have a config management plugin be able to pull in secondary repos with just git clone <repo URL>, as opposed to having to include credentials in plain text.

Yeah, that's what I've been wondering about. The config management plugins are executed in the context of the argocd-repo-server workload, which has no access to Kubernetes API and therefore, not to the Argo CD settings.

**Comment by scalen** [comment]
> The config management plugins are executed in the context of the argocd-repo-server workload, which has no access to Kubernetes API and therefore, not to the Argo CD settings.

Interesting... The repo server has no access to the repo credentials?  To be honest, I haven't tested this with a config management plugin, only with the builtin Helm config manager extended with the Helm-Git plugin.

On that note, though: is the builtin Helm config manager run in a different workload from custom config managers? Because I have used this in conjunction with the Helm-Git plugin to enable specifying chart dependencies in secondary private Git repositories:
- Without this change, the access to secondary repositories fails to be authenticated;
- With this change, the authentication succeeds and the Application can be created.

**Comment by jannfis** [comment]
> On that note, though: is the builtin Helm config manager run in a different workload from custom config managers? Because I have used this in conjunction with the Helm-Git plugin to enable specifying chart dependencies in secondary private Git repositories:

Have you tested this only locally (e.g. `make start`) or actually with a container image built and deployed to a cluster? Our default manifests do not mount the Kubernetes configuration to the argocd-repo-server pod: https://github.com/argoproj/argo-cd/blob/7e68b1ef7157d88049754b4e16209badc5f9f2b0/manifests/base/repo-server/argocd-repo-server-deployment.yaml#L18

We run all configuration tools (e.g. Helm, Kustomize, etc) inside the repository server. The details & credentials for the repositories to be used are being passed in the API call to the repository server. But I'm pretty sure that a call to the Settings API would fail in the repository server, if actually being run within a cluster, because it reads Kubernetes resources. This would however actually work when you run the workloads locally using `make start` or `make start-local`, because your client configuration would be used.

I guess with the new config management plugins concept (#5293, coming most likely with v2.2), users could opt-in for their config management tools to have access to the K8s API, but I wouldn't recommend it for security purposes.

**Comment by scalen** [comment]
> > On that note, though: is the builtin Helm config manager run in a different workload from custom config managers? Because I have used this in conjunction with the Helm-Git plugin to enable specifying chart dependencies in secondary private Git repositories:
> 
> Have you tested this only locally (e.g. `make start`) or actually with a container image built and deployed to a cluster? Our default manifests do not mount the Kubernetes configuration to the argocd-repo-server pod:
> 
> https://github.com/argoproj/argo-cd/blob/7e68b1ef7157d88049754b4e16209badc5f9f2b0/manifests/base/repo-server/argocd-repo-server-deployment.yaml#L18

I tested this in a full container-based deployment to a cluster.  However, I deployed using the [community-maintained Helm chart](https://github.com/argoproj/argo-helm/tree/master/charts/argo-cd), which sets `automountServiceAccountToken: true` by default for the server, controller _and_ repo server.

Could you point me to documentation regarding which components do what? I had assumed that the repo server was responsible for managing the configuration for and interacting with repositories... but if it can't access the repo database, I must be mistaken!

**Comment by scalen** [comment]
Also, @jannfis, I assume that there is some way for the repository credentials can be retrieved by the repo server...  Could you point me at the correct package to hook into? then I can re-work it to retrieve the credentials by that route.

**Comment by jannfis** [comment]
> Could you point me to documentation regarding which components do what? I had assumed that the repo server was responsible for managing the configuration for and interacting with repositories... but if it can't access the repo database, I must be mistaken!

I'm afraid that there is no concrete documentation on that yet. But in a nutshell, `argocd-server` manages configuration, `argocd-application-controller` is responsible for reconciliation and `argocd-repo-server` interacts with repositories and generates the appropriate manifests for a given application. The `argocd-repo-server` is instructed from outside (usually, either by `argocd-application-controller` or the `argocd-server`) to perform actions, and all required configuration (such as the repository configuration and credentials) is supplied from the callers as well.

> Also, @jannfis, I assume that there is some way for the repository credentials can be retrieved by the repo server... Could you point me at the correct package to hook into? then I can re-work it to retrieve the credentials by that route.

The credentials are supplied in the gRPC API call used to request manifest generation. The entry point would be here: https://github.com/argoproj/argo-cd/blob/7e68b1ef7157d88049754b4e16209badc5f9f2b0/reposerver/repository/repository.go#L291

For example, the `argocd-application-controller` calls `GenerateManifest` upon reconciliation here: https://github.com/argoproj/argo-cd/blob/7e68b1ef7157d88049754b4e16209badc5f9f2b0/controller/state.go#L161

> Alternately, this could be left as a feature that is only accessible if the cluster-owners elect to auto mount the ServiceAccountToken on the repo server 🤷🏼

That's certainly an option, but I really think people should not opt-in to do it and I also wonder why the Helm chart maintainers have chosen to make it the default. The Argo CD repo server executes third party binaries, and imho those should be executed as isolated as possible.

**Comment by scalen** [comment]
Firstly, thanks for all the information!

I think the most important line for the future of this PR is the last one:

> The Argo CD repo server executes third party binaries, and imho those should be executed as isolated as possible.

With that principle in mind, the repo server should never be able to _request_ credentials for a repository, only be given them by an incoming gRPC call.  That fundamentally makes the principle of this PR untenable, as it explicitly involves requesting the credentials for secondary repositories (i.e. repositories other than that whose credentials were passed via the gRPC GenerateManifest request).

Given that the functionality is antithetical to the full-security form of deployment, I would suggest that there is no point in trying to develop a solution that will work in that case;  As such, we are left with the options of:
1. Close this PR, so as to not encourage users to run their installations with reduced security.
2. Approve the PR, but as functionality only available when running with reduced security.  I can add documentation of the feature, with a warning about what configuration would need changing to make it work, and to be extra vigilant in that case about what other software is installed on the repo-server.

**Comment by jannfis** [comment]
Despite all my concerns I came up with during our chat, I still think the use case described in #6591 is very valid. 

We already pass credentials for configured (and allowed) Helm repositories when calling the repository server's API in order to allow cross-repository dependencies to work.

There is also an open enhancement request at https://github.com/argoproj/argo-cd/issues/6440 which is about doing a similar thing for Git repositories to allow cross-repository Kustomize remote bases to work.

So, I think it may be viable to sort out the required changes and possibly also pass down configured Git credentials (if allowed by the AppProject) to the repository server and then pick them up from a credential helper for helm+git to work. I think this will require some design work beforehand, but would probably a great feature to have.

**Comment by alexey907** [comment]
> There is also an open enhancement request at #6440 which is about doing a similar thing for Git repositories to allow cross-repository Kustomize remote bases to work.

HI @jannfis 

Thanks for bringing my feature request here. 

Just to avoid confusion: my ticket suggests using git credentials helper of type "store", but that's not the main point of this feature requests, but just a suggestion how to implement it. 

We can use whatever git credentials helper binary, the only thing I care about in that ticket is that it will provide credentials for every git repo listed as "sourceRepos" in my AppProj for the given Application, not just the one referenced in the Application ".source.repoURL" attribute.

**Comment by scalen** [comment]
@alexey907 cool, thanks for the clarification. Yeah, the key feature for both the Kustomize and Helm use cases is getting the credentials exposed to the on host Git process, and the key functional requirement/restriction seems to be that no credentials be accessible on the repo server _outside the context of a sync_.

My understanding of the above suggests that we need to:
1. Keep the credentials within the memory of the repo server's process.
2. Retrieve the credentials through a Git credential helper.

Both my solution of calling out to the source of truth, and @alexey907 solution of writing to a file, violate 1.

My first thought is to add a socket file for the repo server to serve any credentials it has been sent, then have the Git credential helper send requests to the socket. The credential helper will get no results between syncs, and only those sent as part of the request otherwise.

One thing I don't know is if the repo server can handle more than one simultaneous sync... If so, we could avoid cross-provision of credentials by somehow including the request ID in the call to the credential helper... Not sure how to do that at the moment though 😛

**Review by todaywasawesome — CHANGES_REQUESTED** [review]
I'm going to close this issue as it hasn't been updated in 2 years. If it's still relevant, please reopen.

**Comment by adam63redhat** [comment]
> I'm going to close this issue as it hasn't been updated in 2 years. If it's still relevant, please reopen.

@todaywasawesome -- I believe this is still relevant. I'm still not able to use private remote bases with kustomize

**Comment by todaywasawesome** [comment]
@adam63redhat agreed. Reopening.

Another related proposal:
https://github.com/argoproj/argo-cd/issues/15925

**Comment by crenshaw-dev** [comment]
> I'm still not able to use private remote bases with kustomize

I'm surprised by that... I thought we'd built a GIT_ASKPASS binary to facilitate private remote bases.

**Comment by adam63redhat** [comment]
> > I'm still not able to use private remote bases with kustomize
> 
> I'm surprised by that... I thought we'd built a GIT_ASKPASS binary to facilitate private remote bases.
> 
> As a matter of fact, I'm pretty sure Intuit uses this feature.

It may be operator error on my end, but I've not been able to figure it out. I'm working with ADO and ssh, I haven't tried https.

I saw this open issue that looks related: #8342

**Comment by adam63redhat** [comment]
> > > I'm still not able to use private remote bases with kustomize
> > 
> > 
> > I'm surprised by that... I thought we'd built a GIT_ASKPASS binary to facilitate private remote bases.
> > As a matter of fact, I'm pretty sure Intuit uses this feature.
> 
> It may be operator error on my end, but I've not been able to figure it out. I'm working with ADO and ssh, I haven't tried https.
> 
> I saw this open issue that looks related: #8342

Indeed it was operator error: ssh.dev.azure.com has a bunch of entries I needed to add to known_hosts. After doing so, this is working for me as expected.

**Comment by todaywasawesome** [comment]
Based on the resolution I think the issue is closed and this PR is no longer relevant. Closing.

**Pull request closed without merging**